### PR TITLE
Add tiered caching blog

### LINKED ISSUE #3374: [BLOG] Tiered caching in OpenSearch
### Describe the blog post

This will cover high level overview of Tiered caching in OpenSearch, its benefits and how can users use it. It will also includes details like in what cases it makes sense to use it and what are the things user need to keep in mind.

### Expected Title

Tiered caching in OpenSearch

### Authors Name
Sagar Upadhyaya
Ankit Jain
Peter Alfonsi
Kunal Khatua
Fanit Kolchina

### Authors Email
sagar.upadhyaya.121@gmail.com

### Target Draft Date

10/11/2024

### Blog Post Category

technical

### Target Publication Date

10/17/2024

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by kolchfa-aws**
Closes #3374 

### Check List
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

### CODE REVIEW

File: _posts/2024-10-11-tiered-cache.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+   - upasagar
+   - akjain
+   - kkhatua
+   - kolchfa
+date: 2024-10-11
+categories:
+  - technical-posts
+has_science_table: true
+meta_keywords: OpenSearch, tiered caching, query performance, on-heap cache, disk-based cache, request cache, cache hit ratio, OpenSearch Benchmark, performance optimization, caching
+meta_description: Learn how tiered caching in OpenSearch enhances query performance by combining on-heap and disk-based caching. Explore its benefits, limitations, and practical use cases in this detailed blog.
```
Comment (pajuric) [inline _posts/2024-10-11-tiered-cache.md:14]: Updated meta: 
_meta_keywords:_ tiered caching, disk-based caching, on-heap caching, OpenSearch caching performance, how tiered caching works 

_meta_description:_ Explore how OpenSearch combines on-heap and disk-based caching to handle larger datasets and improve performance. Learn about the trade-offs of tiered caching, how it works, and future developments.

File: _posts/2024-10-11-tiered-cache.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+![Performance test results](/assets/media/blog-images/2024-10-11-tiered-cache/performance_results.png){:class="img-centered"}
+
+Initial results show that tiered caching performs well, especially with higher cache hit ratios and latencies below p75. The gains are particularly notable in running computationally expensive queries, because tiered caching reduces the need to recompute them, fetching results directly from the cache instead.
+
+## What’s next?
+
+While tiered caching is a promising feature, we're actively working on further improvements. We're currently exploring ways to make tiered caching more performant. Future enhancements may include promoting frequently accessed items from the disk cache to the on-heap cache, persisting disk cache data between restarts, or integrating tiered caching with other OpenSearch cache types, such as the query cache. You can follow our progress in [this issue](link). We encourage you to try tiered caching in a non-production environment and share any feedback to help us make this feature more robust.
```
Comment (natebower) [inline _posts/2024-10-11-tiered-cache.md:92]:
```suggestion
While tiered caching is a promising feature, we're actively working on further improvements. We're currently exploring ways to make tiered caching more performant. Future enhancements may include promoting frequently accessed items from the disk cache to the on-heap cache, persisting disk cache data between restarts, or integrating tiered caching with other OpenSearch cache types, such as the query cache. You can follow our progress in [this issue](link). We encourage you to try tiered caching in a non-production environment and to share your feedback to help make this feature more robust.
```

**Review by natebower — COMMENTED** [review]
@kolchfa-aws @pajuric Editorial review complete. Please see my changes and let me know if you have any questions. Thanks!

File: _posts/2024-10-11-tiered-cache.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+## Is there a better way?
+
+As discussed, on-heap caches have limitations when handling larger datasets. A more effective caching mechanism is *tiered caching*, which uses multiple cache layers, starting with on-heap caching and extending to a disk-based tier. This approach balances performance and capacity, allowing you to store larger datasets without consuming valuable heap memory.
+
+In the past, using a disk for caching raised concerns because traditional spinning hard drives were slower. However, advancements in storage technology, like modern SSD and NVMe drives, now deliver much faster performance. Although disk access is still slower than memory, the speed gap has narrowed enough that the performance trade-off is minimal and often outweighed by the advantage of increased storage capacity.
```
Comment (kolchfa-aws) [inline _posts/2024-10-11-tiered-cache.md:29]:
```suggestion
In the past, using a disk for caching raised concerns because traditional spinning hard drives were slower. However, advancements in storage technology, like modern SSD and NVMe drives, now deliver much faster performance. Although disk access is still slower than memory, the speed gap has narrowed enough that the performance trade-off is minimal and often outweighed by the benefit of increased storage capacity.
```

File: _posts/2024-10-11-tiered-cache.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+![Performance test results](/assets/media/blog-images/2024-10-11-tiered-cache/performance_results.png){:class="img-centered"}
+
+The initial results show that tiered caching performs well, especially with higher cache hit ratios and latencies lower than p75. The gains are particularly notable when running computationally expensive queries because tiered caching reduces the need to recompute them, fetching results directly from the cache instead.
+
+## What's next?
+
+While tiered caching is a promising feature, we're actively working on further improvements. We're currently exploring ways to make tiered caching more performant. Future enhancements may include promoting frequently accessed items from the disk cache to the on-heap cache, persisting disk cache data between restarts, or integrating tiered caching with other OpenSearch cache types, such as the query cache. You can follow our progress in [this issue](link). We encourage you to try tiered caching in a non-production environment and to share your feedback to help make this feature more robust.
```
Comment (kolchfa-aws) [inline _posts/2024-10-11-tiered-cache.md:92]:
```suggestion
While tiered caching is a promising feature, we're actively working on further improvements. We're currently exploring ways to make tiered caching more performant. Future enhancements may include promoting frequently accessed items from the disk cache to the on-heap cache, persisting disk cache data between restarts, or integrating tiered caching with other OpenSearch cache types, such as the query cache. You can follow our progress in [this issue](https://github.com/opensearch-project/OpenSearch/issues/10024). We encourage you to try tiered caching in a non-production environment and to share your feedback to help make this feature more robust.
```

**Comment by kolchfa-aws** [comment]
@pajuric Editorial comments addressed. This blog is ready to publish.

**Comment by sgup432** [comment]
@kolchfa-aws 
> Editorial comments addressed. This blog is ready to publish.

We still need to add Peter before we publish this blog, right?

**Comment by kolchfa-aws** [comment]
Yes, I am waiting for his info and will add it to this PR when it's available.

**Comment by kolchfa-aws** [comment]
@sgup432 @peteralfonsi Peter's bio added.

File: _posts/2024-10-11-tiered-cache.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+has_science_table: true
+meta_keywords: tiered caching, disk-based caching, on-heap caching, OpenSearch caching performance, how tiered caching works
+meta_description: Explore how OpenSearch combines on-heap and disk-based caching to handle larger datasets and improve performance. Learn about the trade-offs of tiered caching, how it works, and future developments.
+---
+
+For performance-intensive applications like OpenSearch, caching is an essential optimization. Caching stores data so that future requests can be served faster, improving query response times and application performance. OpenSearch uses two main cache types: a request cache and a query cache. Both are on-heap caches, meaning their size is determined by the amount of available heap memory on a node.
+
+## On-heap caching: A good start, but is it enough?
+
+On-heap caching in OpenSearch provides a quick, simple, and efficient way to cache data locally on a node. It offers low-latency data retrieval and thereby provides significant performance gains. However, these advantages come with trade-offs, especially as the cache grows, which may lead to performance challenges.
```
Comment (jainankitk) [inline _posts/2024-10-11-tiered-cache.md:22]: > However, these advantages come with trade-offs, especially as the cache grows, which may lead to performance challenges.

Maybe it is just me, but this line reads slightly incomplete or looks confusing

File: _posts/2024-10-11-tiered-cache.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+Tiered caching is also designed to be pluggable. You can seamlessly integrate different types of on-heap and disk cache implementations or libraries using tiered cache settings. For more information, see [Tiered cache](https://opensearch.org/docs/latest/search-plugins/caching/tiered-cache/).
+
+## When to use tiered caching
+
+Because tiered caching currently only applies to the request cache, it's useful when the existing on-heap request cache isn't large enough to store your datasets and you encounter frequent evictions. You can check request cache statistics using the `GET /_nodes/stats/indices/request_cache` endpoint to monitor evictions, hits, and misses. If you notice frequent evictions along with some hits, enabling tiered caching could provide a significant performance boost.
```
Comment (jainankitk) [inline _posts/2024-10-11-tiered-cache.md:48]: can we roughly quantify `along with some hints`?

File: _posts/2024-10-11-tiered-cache.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+Tiered caching is especially beneficial in these situations:
+
+- Your domain experiences many cache evictions and has repeatable queries. You can confirm this by using request cache statistics.
```
Comment (jainankitk) [inline _posts/2024-10-11-tiered-cache.md:52]: nit: repeatable -> repeating/repeated ?

File: _posts/2024-10-11-tiered-cache.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+- You're working with log analytics or read-only indexes, in which data doesn't change often, and you're encountering frequent evictions.
```
Comment (jainankitk) [inline _posts/2024-10-11-tiered-cache.md:53]: nit: indexes -> indices

> in which data doesn't change often

looks redundant I guess. read-only indices is self explanatory

File: _posts/2024-10-11-tiered-cache.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+By default, the request cache only stores aggregation queries. You can enable caching for specific requests by using the `?request_cache=true` query parameter.
```
Comment (jainankitk) [inline _posts/2024-10-11-tiered-cache.md:55]: > stores aggregation queries

stores aggregation query results

File: _posts/2024-10-11-tiered-cache.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+![Performance test results](/assets/media/blog-images/2024-10-11-tiered-cache/performance_results.png){:class="img-centered"}
+
+The initial results show that tiered caching performs well, especially with higher cache hit ratios and latencies lower than p75. The gains are particularly notable when running computationally expensive queries because tiered caching reduces the need to recompute them, fetching results directly from the cache instead.
+
+## What's next?
+
+While tiered caching is a promising feature, we're actively working on further improvements. We're currently exploring ways to make tiered caching more performant. Future enhancements may include promoting frequently accessed items from the disk cache to the on-heap cache, persisting disk cache data between restarts, or integrating tiered caching with other OpenSearch cache types, such as the query cache. You can follow our progress in [this issue](https://github.com/opensearch-project/OpenSearch/issues/10024). We encourage you to try tiered caching in a non-production environment and to share your feedback to help make this feature more robust.
```
Comment (jainankitk) [inline _posts/2024-10-11-tiered-cache.md:93]: > share your feedback to help make this feature more robust

share your feedback to help improve this feature

Comment (jainankitk) [inline _posts/2024-10-11-tiered-cache.md:93]: > restarts, or integrating

restarts, and integrating

**Review by jainankitk — COMMENTED** [review]
Nice blog @sgup432! Minor comments to help improve the readability

File: _posts/2024-10-11-tiered-cache.md (hunk 5 above)
Comment (kolchfa-aws) [reply _posts/2024-10-11-tiered-cache.md:22]: Reworded.

File: _posts/2024-10-11-tiered-cache.md (hunk 8 above)
Comment (kolchfa-aws) [reply _posts/2024-10-11-tiered-cache.md:53]: Indexes is used per our style guide.

File: _posts/2024-10-11-tiered-cache.md [hunk 11]
```diff
@@ -19,7 +19,7 @@ For performance-intensive applications like OpenSearch, caching is an essential
 
 ## On-heap caching: A good start, but is it enough?
 
-On-heap caching in OpenSearch provides a quick, simple, and efficient way to cache data locally on a node. It offers low-latency data retrieval and thereby provides significant performance gains. However, these advantages come with trade-offs, especially as the cache grows, which may lead to performance challenges.
+On-heap caching in OpenSearch provides a quick, simple, and efficient way to cache data locally on a node. It offers low-latency data retrieval and thereby provides significant performance gains. However, these advantages come with trade-offs: as the cache grows, on-heap caching may lead to performance challenges.
```
Comment (natebower) [inline _posts/2024-10-11-tiered-cache.md:22]: Capitalize "As" after the colon.

**Review by natebower — COMMENTED** [review]
@kolchfa-aws Just one minor change.

File: _posts/2024-10-11-tiered-cache.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+has_science_table: true
+meta_keywords: tiered caching, disk-based caching, on-heap caching, OpenSearch caching performance, how tiered caching works
+meta_description: Explore how OpenSearch combines on-heap and disk-based caching to handle larger datasets and improve performance. Learn about the trade-offs of tiered caching, how it works, and future developments.
+---
+
+For performance-intensive applications like OpenSearch, caching is an essential optimization. Caching stores data so that future requests can be served faster, improving query response times and application performance. OpenSearch uses two main cache types: a request cache and a query cache. Both are on-heap caches, meaning their size is determined by the amount of available heap memory on a node.
+
+## On-heap caching: A good start, but is it enough?
+
+On-heap caching in OpenSearch provides a quick, simple, and efficient way to cache data locally on a node. It offers low-latency data retrieval and thereby provides significant performance gains. However, these advantages come with trade-offs: as the cache grows, on-heap caching may lead to performance challenges.
```
Comment (kolchfa-aws) [inline _posts/2024-10-11-tiered-cache.md:22]:
```suggestion
On-heap caching in OpenSearch provides a quick, simple, and efficient way to cache data locally on a node. It offers low-latency data retrieval and thereby provides significant performance gains. However, these advantages come with trade-offs: As the cache grows, on-heap caching may lead to performance challenges.
```

File: _posts/2024-10-11-tiered-cache.md (hunk 6 above)
Comment (sgup432) [reply _posts/2024-10-11-tiered-cache.md:48]: @kolchfa-aws @jainankitk I think we can leave it as it is. I deliberately left it like that i.e. not quantify it as it is hard to say whether tiered cache will only benefit with >50% or >30% cache hit ratio.

File: _posts/2024-10-11-tiered-cache.md (hunk 5 above)
Comment (sgup432) [reply _posts/2024-10-11-tiered-cache.md]: @kolchfa-aws Can we reword this to something like below? As it is more specific.

"However, these advantages come with trade-offs, especially as the cache grows in size and reaches its capacity, which may lead to performance challenges due to high evictions and misses."

File: _posts/2024-10-11-tiered-cache.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+   - peteral
+   - kkhatua
+   - kolchfa
+date: 2024-10-11
+categories:
+  - technical-posts
+has_science_table: true
+meta_keywords: tiered caching, disk-based caching, on-heap caching, OpenSearch caching performance, how tiered caching works
+meta_description: Explore how OpenSearch combines on-heap and disk-based caching to handle larger datasets and improve performance. Learn about the trade-offs of tiered caching, how it works, and future developments.
+---
```
Comment (pajuric) [inline _posts/2024-10-11-tiered-cache.md:16]: Please update the meta with the following: 

meta_keywords:  tiered caching, on-heap cache, disk-based caching, how tiered caching works, OpenSearch cache optimization

meta_description:  Explore the benefits of combining on-heap and disk-based caching in OpenSearch to manage large datasets. Learn how tiered caching works, when to use it, and the performance results of our testing.

File: _posts/2024-10-11-tiered-cache.md [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+layout: post
+title:  "Tiered caching in OpenSearch"
+authors:
+   - upasagar
+   - akjain
+   - peteral
+   - kkhatua
+   - kolchfa
+date: 2024-10-11
```
Comment (pajuric) [inline _posts/2024-10-11-tiered-cache.md:10]: Update blog date to 2024-10-24

**Comment by pajuric** [comment]
@kkhatua - You are approved to push this live.

**Review by peterzhuamazon — APPROVED** [review]
Push to staging.
